src/main/java/org/embulk/input/marketo/delegate/MarketoBaseBulkExtractInputPlugin.java

```diff
@@ -0,0 +1,199 @@
+        super.validateInputTask(task);
+    }
+
+    @Override
+    public ConfigDiff buildConfigDiff(T task, Schema schema, int taskCount, List<TaskReport> taskReports)
+    {
+        ConfigDiff configDiff = super.buildConfigDiff(task, schema, taskCount, taskReports);
+        Long currentLatestFetchTime = 0L;
+        Set<String> latestUIds = null;
+        for (TaskReport taskReport : taskReports) {
+            Long latestFetchTime = taskReport.get(Long.class, LATEST_FETCH_TIME);
+            if (latestFetchTime == null) {
+                continue;
+            }
+            if (currentLatestFetchTime < latestFetchTime) {
+                currentLatestFetchTime = latestFetchTime;
+                latestUIds = taskReport.get(Set.class, LATEST_UID_LIST);
+            }
+        }
+        if (!currentLatestFetchTime.equals(0L)) {
+            configDiff.set(LATEST_FETCH_TIME, currentLatestFetchTime);
+            DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ");
+            configDiff.set("from_date", df.format(new Date(currentLatestFetchTime)));
+            configDiff.set(LATEST_UID_LIST, latestUIds);
+        }
+        return configDiff;
+    }
+
+    @Override
+    public TaskReport ingestServiceData(T task, RecordImporter recordImporter, int taskIndex, PageBuilder pageBuilder)
+    {
+        InputStream extractedStream = getExtractedStream(task, pageBuilder.getSchema());
+        return importRecordFromFile(task, extractedStream, recordImporter, pageBuilder);
+    }
+
+    protected TaskReport importRecordFromFile(T task, InputStream inputStream, RecordImporter recordImporter, PageBuilder pageBuilder)
+    {
+        Set<String> latestUids = task.previousUIds();
+        TaskReport taskReport = Exec.newTaskReport();
+        int imported = 0;
+        Timestamp currentTimestamp = null;
+        if (task.getLatestFetchTime().isPresent()) {
+            currentTimestamp = Timestamp.ofEpochMilli(task.getLatestFetchTime().get());
+        }
+        TimestampParser timestampParser = new TimestampParser(MarketoUtils.ISO_8601_FORMAT, DateTimeZone.UTC);
+        try (LineDecoder lineDecoder = new LineDecoder(new InputStreamFileInput(task.getBufferAllocator(), inputStream), task)) {
+            CsvTokenizer csvTokenizer = new CsvTokenizer(lineDecoder, task);
+            if (!csvTokenizer.nextFile()) {
+                throw new DataException("Can't read extract input stream");
+            }
+            csvTokenizer.nextRecord();
+            List<String> headers = new ArrayList<>();
+            while (csvTokenizer.hasNextColumn()) {
+                headers.add(csvTokenizer.nextColumn());
+            }
+            while (csvTokenizer.nextRecord()) {
+                List<String> values = new ArrayList<>();
+                while (csvTokenizer.hasNextColumn()) {
+                    values.add(csvTokenizer.nextColumnOrNull());
+                }
+                if (imported >= PREVIEW_RECORD_LIMIT && Exec.isPreview()) {
```

Comment:
This seems not the correct position to check for preview record limit as:
- The `imported` doesn't change since the last successful import
- Looped over the row data without doing anything useful

Reply:
`The imported doesn't change since the last successful import` I don't quite understand this